agents/magpie/magpie_agent.py

```diff
@@ -0,0 +1,388 @@
+                bands_seen += 1
+            color_idx = band_idxs[det['band']]
+
+            fp.xs[chan] = x * wafer_scale
+            fp.ys[chan] = y * wafer_scale
+            fp.rots[chan] = polangle
+            fp.cnames[chan] = key
+            fp.eqs[chan] = f"/ + 1 s {key} 2"
+            fp.cmaps[chan] = cmaps[color_idx]
+            fp.templates[chan] = f"template_c{color_idx}_p0"
+
+        return fp
+
+
+
+class MagpieAgent:
+    mask_register = 'AMCc.SmurfProcessor.ChannelMapper.Mask'
+
+    def __init__(self, agent, args):
+        self.agent = agent
+        self._running = False
+
+        self.target_rate = args.target_rate
+        layout = args.layout.lower()
+        if layout == 'grid':
+            self.fp = FocalplaneConfig.grid(
+                args.xdim, args.ydim, ygap=8
+            )
+        elif layout == 'wafer':
+            if args.csv_file is not None:
+                self.fp = FocalplaneConfig.from_tune_csv(
+                    args.csv_file, wafer_scale=args.wafer_scale
+                )
+            else:
+                self.fp = FocalplaneConfig.from_wafer_file(
+                    args.wafer_file, wafer_scale=args.wafer_scale
+                )
+
+        self.sent_cfg = False
+        self.num_dets = self.fp.num_dets
+        self.mask = np.arange(self.num_dets)
+        self.mask = np.arange(4960)
+
+        self.reader = core.G3Reader(args.src)
+        self.sender = core.G3NetworkSender(
+            hostname='*', port=args.dest, max_queue_size=1000
+        )
+
+        self.lowpass = FIRFilter.butter_lowpass(self.target_rate, 200., order=4)
+        self.send_config_flag = True
+        self.scale = 1
+        self.offsets = np.zeros(self.fp.num_dets)
+        self.zero_flag = True
+
+        self.demodulate=True
+        self.demod_filter = FIRFilter.butter_lowpass(1, 200., order=4)
+        self.demod_freq = 4
+
+    def set_scale(self, session, params={}):
+        self.scale = params['scale']
+        return True, f"Set scale to {self.scale}"
+
+    def zero_dets(self, session, params={}):
+        self.zero_flag = True
+        return True, "Zeroing detectors"
+
+    def set_target_rate(self, session, params={}):
+        """
+        Args
+        ----
+        target_rate : float
+            Target sample rate for lyrebird (Hz)
+        """
+        self.target_rate = params['target_rate']
+        return True, f'Set target rate to {self.target_rate}'
+
+    def set_demod(self, session, params={}):
+        if 'state' in params:
+            self.demodulate = params['state']
+        if 'freq' in params:
+            self.demodulate = params['freq']
+        if ('lowpass_order' in params) or ('lowpass_cutoff'):
+            order = params.get('lowpass_order', 4)
+            cutoff = params.get('lowpass_cutoff', 1)
+            self.demod_filter = FIRFilter.butter_lowpass(cutoff, 200., order=order)
+        return True, f"set demod {params}"
+
+    def process_status(self, frame):
```

Comment:
Since "process" now conflicts with OCS "Process" language, it's even more useful for the docstrings to point out which things in here are Task and Process functions... and which are just frame handling helpers.